_posts/2023-04-30-writing-code-with-llm-dont-forget-kani.md

```diff
@@ -0,0 +1,306 @@
+While the generated code is not always sound, program verifiers like Kani can effectively detect bugs in the generated code and help ChatGPT fix them.
+
+## Code Generation using ChatGPT
+We demonstrate the three tasks with two examples.
+Let's start our first task: can we write code using prompt with ChatGPT?
+
+#### Example 1.a
+The first example is a toy problem of writing a rust function to compute the average of two integers.
+So the prompt is also very simple and straightforward.
+
+```
+Q: Can you write me a rust function that compute the integer average of two integers?
+```
+
+Giving the above prompt, ChatGPT answered us with the following function
+```rust
+fn integer_average(a: i32, b: i32) -> i32 {
+    (a + b) / 2
+}
+```
+, the explanation of the function
+>In this function, we define a function called integer_average that takes two integer arguments a and b.
+The function then adds the two arguments together and divides the result by 2 to compute the integer average.
+
+, and even a drive function to call the average function.
+```rust
+fn main() {
+    let a = 4;
+    let b = 8;
+    let result = integer_average(a, b);
+    println!("The average of {} and {} is {}", a, b, result);
+}
+```
+
+#### Example 2.a
+The second example we tried is more like a real-world coding task.
+We wrote a rust function that compute the lowest unset bit in a `u32` integer.
+Our implementation uses a loop to iterate through all 32 bits in input.
+And we know that this could be slower than using some [bit twiddling hacks](https://graphics.stanford.edu/~seander/bithacks.html).
+So, we want to optimize the function with ChatGPT.
+The prompt we provided including the implementation we wrote.
+
+``` 
+Q: Optimize this rust function:
+fn lowest_unset_bit(x: u32) -> Option<u8> {
+    let mut mask = 1;
+    for i in 0..32 {
+        if x & mask == 0 {
+            return Some(i);
+        }
+        mask <<= 1;
+    }
+    None
+} 
+```
+
+After a few seconds, ChatGPT gave us a solution
+``` rust
+fn lowest_unset_bit(x: u32) -> Option<u8> {
+    if x == std::u32::MAX{
+        return None;
+    }
+    let y = x | (x + 1);
+    let z = y ^ (y + 1);
+    Some((z.trailing_zeros() % 32) as u8)
+}
+```
+with the explanation of it.
+
+
+From these two examples, I must say, I'm thoroughly impressed with ChatGPT's code generation capabilities.
+Not only does it produce code that looks good at first glance, but it also provides explanations that give us greater confidence that the code is doing exactly what we intended it to do.
+In fact, we can even request ChatGPT to generate a driver function and some test cases using follow-up prompts to further test the generated code.
+However, as a verification folk, we know that we still can't be completely sure that the code is correct without **verification**.
+
+
+## Verification of Generated Code
+So, naturally, our next task is: can we verify those generated code with Kani?
+
+#### Example 1.b
```

Comment:
These numbers are a bit confusing. I read that this is a different example than `1.a`. It almost read like a typo to me when I move to the next line.